### PR TITLE
Denoiser for plug and play.

### PR DESCRIPTION
**Pull request by paquiteau**
This Pull Request add Denoising operators. 

Implementation-wise they are very similar to Proximal operators, and can typically be used as a drop-in replacement for them in iteratives algorithms making them "plug and play algorithms". 

For now BM3D (via external package) and local low rank denoiser (MP-PCA) are implemented, let me know if you would like to see other methods implemented. 

TODO: 
- [ ] add tests for coverage
- [ ] provides implementation examples

### CODE REVIEW

**Comment by philouc** [comment]
Very nice piece of work. I was wondering if we could try later on to plug the Learnlets denoiser that Zacch developed with Jean-Luc. I'm looking forward to seeing the performance of the local low rank denoiser at work during image recon.

File: modopt/opt/proximity/denoiser.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    BM3D_AVAILABLE = False
+
+
+class BM3Denoise(ProximityParent):
+    """
+    BM3D Denoiser.
+
+    This a wrapper around the BM3D Denoiser, using the implementation of
+    :cite:`makinen2020`.
+    """
```
Comment (sfarrens) [inline modopt/opt/proximity/denoiser.py:26]: Missing parameters section.

File: modopt/opt/proximity/denoiser.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    the MATLAB implementation of  [1]
+
+
+    Parameters
+    ----------
+    patch_size: tuple
+        the patch size to consider for the local low rank regularisation
+    patch_overlap:
+        the amount of pixel/voxel overlap of each patch.
+        If > 0 the patch are averaged.
```
Comment (sfarrens) [inline modopt/opt/proximity/denoiser.py:57]: Parameters section incomplete and it looks like `patch_size` should be `patch_shape`.

File: modopt/opt/proximity/denoiser.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+        if threshold_estimation not in {'MPPCA', 'NORDIC'}:
+            raise ValueError(
+                'Threshold estimation method should be "MPPCA" or "NORDIC"',
+            )
+
+        self.threshold_estimation = threshold_estimation
+        self.noise_weight = noise_weights
+        self.mask = mask
+
+    def _make_patch_shape(self, ndim):
```
Comment (sfarrens) [inline modopt/opt/proximity/denoiser.py:87]: Add docstring even for hidden methods.

File: modopt/opt/proximity/denoiser.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    This a wrapper around the BM3D Denoiser, using the implementation of
+    :cite:`makinen2020`.
+    """
+
+    def __init__(self, noise_std=1.0):
+        self.noise_std = noise_std
+        self.op = self._op_method
+        self.cost = lambda x_val: 0
+
+    def _op_method(self, input_data, extra_factor=1.0):
```
Comment (sfarrens) [inline modopt/opt/proximity/denoiser.py:33]: Add docstring even for hidden methods.

File: modopt/opt/proximity/denoiser.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+                    *((1,) * (len(self.patch_shape) - ndim)),
+                )
+            else:
+                patch_shape = self.patch_shape
+
+        elif isinstance(self.patch_shape, int):
+            patch_shape = (self.patch_shape,) * ndim
+        return patch_shape, patch_overlap
+
+    def _op_method(self, input_data, extra_factor=None):
```
Comment (sfarrens) [inline modopt/opt/proximity/denoiser.py:117]: Add docstring even for hidden methods.

File: modopt/opt/proximity/denoiser.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
+"""Denoiser Operators.
+
+Denoising operators that can be used as a drop-in replacement of proximal
+operators in iterative algorithms, making them a "plug-and-play" method.
+
+TODO Add support for VBM4D (video denoising, may be usefull for dynamic data).
+TODO Add support for Local Low-Rank Denoising.
```
Comment (sfarrens) [inline modopt/opt/proximity/denoiser.py:7]: Do you want this to be rendered in the HTML?

**Comment by paquiteau** [comment]
So I think this PR is becoming too huge to be reasonably reviewed and added to ModOpt (the Local low rank (LLR) denoising methods are over 500 LoC, and yet without test or example)

After thinking about it we have several choices:
1) Do nothing, process the PR as it is (after adding tests and examples)
2) Create a new branch and PR for the LLR methods and do the same process there
3) Extract all of the code of LLR and create a dedicated python module (ie make a new repo, under paquiteau/ or CEA-COSMIC/), and make it an optional dependencies of ModOpt (like for BM3D). This has the benefit to make the LLR denoising methods available to everyone without necessarly install ModOpt to use it, and will makes its development easier for me. However I won't get the review process like here.

**Comment by sfarrens** [comment]
Hi @paquiteau I see your point. Personally I would prefer option 2 as this will be easier to review and keeps everything under the umbrella of ModOpt, which means it will be maintained. Option 1 is also manageable, but will certainly take longer for the PR to be merged. I am not against option 3, but it would probably be best to discuss this at the next PySAP meeting to see what the team thinks.

**Pull request closed without merging**